Console/Command/BatchingOptimizeCommand.php

```diff
@@ -0,0 +1,393 @@
+            0;
+
+        $this->storeCounts[$storeId]['complex_percentage'] = $this->storeCounts[$storeId]['total'] > 0 ?
+            ($this->storeCounts[$storeId]['complex'] * 100) / $this->storeCounts[$storeId]['total']:
+            0;
+
+
+        $sampleSize = $this->input->getOption(self::OPTION_SAMPLE_SIZE) ?? self::DEFAULT_SAMPLE_SIZE;
+        $simpleSampleSize = (int)round($sampleSize * ($this->storeCounts[$storeId]['simple_percentage'] / 100));
+        $complexSampleSize = (int)round($sampleSize * ($this->storeCounts[$storeId]['complex_percentage'] / 100));
+
+        $this->storeCounts[$storeId]['simple_sample_size'] = $simpleSampleSize;
+        $this->storeCounts[$storeId]['complex_sample_size'] = $complexSampleSize;
+
+        $this->storeCounts[$storeId]['sample'] = array_merge(
+            $this->getProductsSizes($simpleProducts, $simpleSampleSize),
+            $this->getProductsSizes($complexProducts, $complexSampleSize)
+        );
+
+        $this->storeCounts[$storeId]['sample_min'] = min($this->storeCounts[$storeId]['sample']);
+        $this->storeCounts[$storeId]['sample_max'] = max($this->storeCounts[$storeId]['sample']);
+    }
+
+    /**
+     * @param int $storeId
+     * @param array $productTypes
+     * @return Collection
+     */
+    protected function getProductsCollectionForStore(int $storeId, array $productTypes = []): Collection
+    {
+        $onlyVisible = !$this->configHelper->includeNonVisibleProductsInIndex();
+        $collection = $this->productHelper->getProductCollectionQuery($storeId, null, $onlyVisible);
+        if (count($productTypes) > 0) {
+            $collection->addAttributeToFilter('type_id', ['in' => $productTypes]);
+        }
+
+        // Randomize the results to get a more "diverse" sample
+        $collection->getSelect()->orderRand();
+
+        return $collection;
+    }
+
+    /**
+     * @param Collection $products
+     * @param int $sampleSize
+     * @return array
+     * @throws LocalizedException
+     * @throws NoSuchEntityException
+     * @throws DiagnosticsException
+     * @throws AlgoliaException
+     */
+    protected function getProductsSizes(Collection $products, int $sampleSize): array
+    {
+        $stats = [];
+        $limit = 0;
+
+        foreach ($products as $product) {
+            if ($limit >= $sampleSize) {
+                break;
+            }
+
+            $serializedRecord = json_encode($this->recordBuilder->buildRecord($product));
+
+            if (function_exists('mb_strlen')) {
+                $size = mb_strlen($serializedRecord, '8bit');
+            } else {
+                $size = strlen($serializedRecord);
+            }
+
+            $stats[$product->getSku()] = $size;
+            $limit++;
+        }
+
+        return $stats;
+    }
+
+    /**
+     * @param array $sizes
+     * @return int
+     */
+    protected function getSizeAverage(array $sizes): int
+    {
+        if (count($sizes) <= 1) {
```

Comment:
```suggestion
        if (empty($sizes)) {
```

That earlier suggestion was intended for the standard deviation `n-1` adjustment which doesn't apply here. Here we can just check for `0`.